hibernate-core/src/main/java/org/hibernate/query/Pager.java

```diff
@@ -0,0 +1,247 @@
+		return query;
+	}
+
+	/**
+	 * @return the current {@linkplain Page page}.
+	 */
+	public Page getCurrentPage() {
+		return page;
+	}
+
+	/**
+	 * @return the current page of results, as a list.
+	 */
+	public List<R> getResultList() {
+		return results;
+	}
+
+	/**
+	 * Advance to the next page of results.
+	 *
+	 * @see Page#next()
+	 */
+	public void next() {
+		page = page.next();
+		execute();
+	}
+
+	/**
+	 * Go back to the previous page of results.
+	 *
+	 * @see Page#previous()
+	 */
+	public void previous() {
+		page = page.previous();
+		if ( lastPage >= 0 ) {
+			hasMoreResults = page.getNumber() <= lastPage;
+		}
+		execute();
+	}
+
+	/**
+	 * @return {@code true} if the {@linkplain #next() next} page has
+	 *		 a nonempty {@linkplain #getResultList() result list}.
+	 *
+	 * @apiNote A return value of {@code true} guarantees that the
+	 *          next page has at least one result. A return value of
+	 *          {@code false} indicates that every subsequent page is
+	 *          empty.
+	 *
+	 * @see #next()
+	 */
+	public boolean hasResultsOnNextPage() {
+		return hasMoreResults;
+	}
+
+	/**
+	 * @return {@code true} unless this is the first page of results.
+	 *
+	 * @apiNote A return value of {@code true} does not guarantee
+	 *          that the {@linkplain #getResultList() result list}
+	 *          of the previous page is nonempty. But a return value
+	 *          of {@code false} does indicate that every earlier
+	 *          page is empty.
+	 *
+	 * @see #previous()
+	 */
+	public boolean hasResultsOnPreviousPage() {
+		return !page.isFirst();
+	}
+
+	/**
+	 * @return {@code true} if the current page has a nonempty
+	 *		 {@linkplain #getResultList() result list}.
+	 */
+	public boolean hasResults() {
+		return !results.isEmpty();
+	}
+
+	/**
+	 * Perform the given action on each remaining page of results,
+	 * starting from the current page.
+	 */
+	public void forEachRemainingPage(Consumer<? super List<R>> action) {
+		new PageIterator().forEachRemaining( action );
+	}
+
+	/**
+	 * Perform the given action on each result, starting from the
+	 * first result of the current page.
+	 */
+	public void forEachRemaining(Consumer<? super R> action) {
+		forEachRemainingPage( page -> page.forEach( action ) );
+	}
+
+	/**
+	 * @return a {@link Stream} over the remaining pages of results,
+	 *		 starting from the current page.
+	 */
+	public Stream<List<R>> stream() {
+		return StreamSupport.stream(
+				spliteratorUnknownSize( new PageIterator(), NONNULL | IMMUTABLE | ORDERED ),
+				false
+		);
+	}
+
+	private void execute() {
+		if ( hasMoreResults ) {
+			// either the previous page was not the last,
+			// or we went backwards using previous(), or
+			// this is the very first time we're executing
+			// the query
+			final int pageSize = page.getMaxResults();
+			// ask for one more row than we actually need,
+			// to detect if there are more results on the
+			// next page
+			query.setMaxResults( pageSize + 1 );
```

Comment:
I'm not sure this feature is a good idea, given that loading one more entity has side effects (persistence context, caches, listeners).

Also not sure how this would play with `fetch/update/flush/clear` loops... probably safe, but who knows.

It's not that hard for users to change their loop so that they stop when a page is empty. That's how the [`Scroll` in Hibernate Search](https://docs.jboss.org/hibernate/search/6.2/reference/en-US/html_single/#search-dsl-query-fetching-results-scrolling) (a paginated cursor) works. However, it kinda requires to present the results as a custom class (`SearchScrollResult`) instead of a `List`, otherwise it's a bit clunky.

Reply:
One of the issues I'm trying to address here is to let people easily run a query for just one page, and then present to the UI whether there are more results or not. (Typical next/last navigation arrows for example.)

I admit this API is trying to do maybe too many things at once, which is why I'm kind of having second thoughts about it.